src/thread.rs

```diff
@@ -0,0 +1,134 @@
+//! Widget that separates resize+encode from rendering.
+//! This allows for rendering to be non-blocking, offloading resize+encode into another thread.
+//! See examples/async.rs for how to setup the threads and channels.
+//! At least one worker thread for resize+encode is required, the example shows how to combine
+//! the needs-resize-polling with other terminal events into one event loop.
+
+use std::sync::mpsc::Sender;
+
+use ratatui::{
+    prelude::{Buffer, Rect},
+    widgets::StatefulWidget,
+};
+
+use crate::{errors::Errors, protocol::StatefulProtocol, Resize};
+
+/// A widget that uses a custom ThreadProtocol as state to offload resizing and encoding to a
+/// background thread.
+pub struct ThreadImage {
+    resize: Resize,
+}
+
+impl ThreadImage {
+    pub const fn resize(self, resize: Resize) -> Self {
+        Self { resize }
+    }
+
+    pub const fn new() -> Self {
+        Self {
+            resize: Resize::Fit(None),
+        }
+    }
+}
+
+impl Default for ThreadImage {
+    fn default() -> Self {
+        Self {
+            resize: Resize::Fit(None),
+        }
+    }
+}
+
+impl StatefulWidget for ThreadImage {
+    type State = ThreadProtocol;
+
+    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
+        if let Some(mut protocol) = state.inner.take() {
+            // If it needs resizing (grow or shrink) then send it away instead of rendering.
+            // Send the requested area instead of the calculated area
+            // to ensure consistent calculations between the render thread and the UI thread.
+            if let Some(area) = protocol.needs_resize(&self.resize, area) {
+                state.increment_id();
+                let _ = state.tx.send(ResizeRequest {
+                    protocol,
+                    resize: self.resize,
+                    area,
+                    id: state.id,
+                });
+            } else {
+                protocol.render(area, buf);
+                state.inner = Some(protocol)
+            }
+        }
+    }
+}
+
+/// The only usage of this struct is to call `perform()` on it and pass the completed resize to `ThreadProtocols` `update_protocol()`
+pub struct ResizeRequest {
+    protocol: StatefulProtocol,
+    resize: Resize,
+    area: Rect,
+    id: u64,
+}
+
+impl ResizeRequest {
+    pub fn perform(mut self) -> Result<CompletedResize, Errors> {
+        self.protocol.resize_encode(&self.resize, self.area);
+        self.protocol
+            .last_encoding_result()
+            .expect("The resize has just been performed")?;
+        Ok(CompletedResize {
+            protocol: self.protocol,
+            id: self.id,
+        })
+    }
+}
+
+/// The only usage of this struct is to pass it to `ThreadProtocols` `update_protocol()`
+pub struct CompletedResize {
```

Comment:
can you rename this to `ResizeResponse`? To match the send type name more closely.